### PR TITLE
[BEAM-551] PubsubIO with ValueProvider

### PR DESCRIPTION
**Pull request by sammcveety**
R: @davorbonaci

### CODE REVIEW

**Comment by sammcveety** [comment]
I'm gonna rebase this on the NVP change now that it's in; any other things I should change along the way?

**Comment by sammcveety** [comment]
Rebase done.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubIO.java [hunk 1]
```diff
@@ -253,6 +257,28 @@ public String asPath() {
   }
 
   /**
+   * Simple {@link SerializableFunction} for {@link PubsubTopic}.
+   */
+  public static class TopicTranslator
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubIO.java]: can these be `private` or at least package private?

Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubIO.java]: Esp if visibility is reduced, could improve javadoc to be something like "Used to build a {@link ValueProvider} for {@link PubsubTopic}"

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubUnboundedSource.java [hunk 2]
```diff
@@ -1305,6 +1306,11 @@ public ProjectPath getProject() {
 
   @Nullable
   public TopicPath getTopic() {
+    return topic.get();
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubUnboundedSource.java]: Something seems off here. I know that `topic.get()` cannot return a `null` value. But `topic` is `@Nullable`, so maybe that is where the `null` should come from?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubIO.java [hunk 3]
```diff
@@ -630,9 +658,9 @@ private Bound(String name, PubsubSubscription subscription, PubsubTopic topic,
                       .setCoder(coder);
         } else {
           @Nullable ProjectPath projectPath =
-              topic == null ? null : PubsubClient.projectPathFromId(topic.project);
-          @Nullable TopicPath topicPath =
-              topic == null ? null : PubsubClient.topicPathFromName(topic.project, topic.topic);
+              topic == null ? null : PubsubClient.projectPathFromId(topic.get().project);
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubIO.java]: I would expect to see no calls to `.get()` in apply for anything that is a VP. Presumably everything that would access `.get()` needs to be a nested VP.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubIO.java [hunk 4]
```diff
@@ -648,7 +676,9 @@ private Bound(String name, PubsubSubscription subscription, PubsubTopic topic,
       @Override
       public void populateDisplayData(DisplayData.Builder builder) {
         super.populateDisplayData(builder);
-        populateCommonDisplayData(builder, timestampLabel, idLabel, topic);
+        PubsubTopic topicString =
+            topic == null ? null : topic.get();
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubIO.java]: same point here as above, w.r.t. `topic.get`. Need also to check `topic.isAccessible()`.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubIO.java (hunk 1 above)
Comment (sammcveety) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubIO.java]: Done

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubIO.java (hunk 3 above)
Comment (sammcveety) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubIO.java]: Yep.  Was checking that we liked the general NVP pattern first, now fixing this.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubIO.java (hunk 4 above)
Comment (sammcveety) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubIO.java]: Done

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubIO.java [hunk 5]
```diff
@@ -508,7 +545,7 @@ private Bound(Coder<T> coder) {
         this(null, null, null, null, coder, null, 0, null);
       }
 
-      private Bound(String name, PubsubSubscription subscription, PubsubTopic topic,
+      private Bound(String name, PubsubSubscription subscription, ValueProvider<PubsubTopic> topic,
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubIO.java]: Doesn't `subscription` also need to be a VP?

**Comment by sammcveety** [comment]
I was thinking that I'd tackle the translator separately, after closing the (still open) question of propagating property names through NVPs, and use the custom IO flavors in the interim.

**Comment by davorbonaci** [comment]
All code changes look great to me!

We could try to do more testing-wise, I think. All test changes seem to be fixups, without any new test cases and some test cases removed. Perhaps we can unit test builders with this new functionality, as there's non trivial logic around when objects may or may not be null.

**Comment by sammcveety** [comment]
Added a test for the builders for Read, and cleaned up the code a bit for Write (since topic should never be null there).

**Comment by dhalperi** [comment]
Merged, thanks.

**Pull request closed without merging**